src/rust/ide/src/controller/searcher/action.rs

```diff
@@ -198,26 +291,97 @@ impl List {
         existing_ids.filter_map(move |id| self.entries.borrow().get(id).cloned())
     }
 
-    /// Extend the list with new actions.
-    ///
-    /// The new actions will be put at end, regardless the current filtering. This function
-    /// is meant to be a part of list's initialization.
-    pub fn extend<T:IntoIterator<Item=Action>>(&self, iter: T) {
-        self.entries.borrow_mut().extend(iter.into_iter().map(ListEntry::from));
-        self.matching.set(0..self.len())
-    }
-
     /// Convert to the action vector.
     ///
     /// Used for testing.
     pub fn to_action_vec(&self) -> Vec<Action> {
         self.entries.borrow().iter().map(|entry| entry.action.clone_ref()).collect()
     }
+
+    fn update_sorting(&self) {
+        let mut entries_mut = self.entries.borrow_mut();
+        entries_mut.sort_by(|l,r| l.ordering_on_list(r));
+        let first_not_matching = entries_mut.partition_point(|e| e.matches());
+        self.matching.set(0..first_not_matching);
+    }
 }
 
-impl<IntoIter> From<IntoIter> for List
-where IntoIter : IntoIterator<Item=Action> {
-    fn from(actions:IntoIter) -> Self {
-        Self::from_actions(actions)
+
+
+// ===================
+// === ListBuilder ===
+// ===================
+
+/// The List Builder.
+///
+/// As the [`List`] itself keeps its categories immutable, the builder is used to extend the list.
+/// It ensures, that the all indexes will remain intact and in a proper order (see also fields docs
+/// in [`List`].
+#[derive(Debug,Default)]
+pub struct ListBuilder {
```

Comment:
Can extend comment that it prevents adding (building) a new subcategory if previously edited is not finished.